### PR TITLE
WIP ext: loading of shared tasks via entry_points

### PR DESCRIPTION
**Pull request by bruno314**
work in progress

### CODE REVIEW

File: invenio_celery/ext.py [hunk 1]
```diff
@@ -44,6 +51,15 @@ def init_app(self, app, assets=None, **kwargs):
         """Initialize application object."""
         self.init_config(app.config)
         self.celery = FlaskCeleryExt(app).celery
+
+        all_modules = []
+        for item in pkg_resources.\
+                iter_entry_points(group=self.entry_point_group):
```
Comment (jirikuncar) [inline invenio_celery/ext.py]: break on opening `(`

File: invenio_celery/ext.py [hunk 2]
```diff
@@ -34,9 +35,15 @@
 class InvenioCelery(object):
     """Invenio theme extension."""
 
-    def __init__(self, app=None, **kwargs):
+    def __init__(self, app=None, entry_point_group="invenio_celery.tasks",
```
Comment (jirikuncar) [inline invenio_celery/ext.py]: move to `entry_point_group="invenio_celery.tasks",` to `init_app`

File: invenio_celery/ext.py [hunk 3]
```diff
@@ -34,9 +35,15 @@
...
+                 **kwargs):
         """Extension initialization."""
         self.celery = None
+        self.entry_point_group = entry_point_group
```
Comment (jirikuncar) [inline invenio_celery/ext.py]: you don't need to store it if you use it only in `init_app`

File: invenio_celery/ext.py [hunk 4]
```diff
@@ -34,9 +35,15 @@
...
+
+        # this reduces signature complexity, we want False only for testing
+        self.enable_autodiscovery = kwargs.get("enable_autodiscovery", True)
```
Comment (jirikuncar) [inline invenio_celery/ext.py]: like https://github.com/inveniosoftware/invenio-celery/pull/10/files#r43890452

File: invenio_celery/ext.py (hunk 2 above)
Comment (jirikuncar) [inline invenio_celery/ext.py]: please follow the naming `entrypoint_name ` see https://github.com/inveniosoftware/invenio-db/blob/master/invenio_db/core.py#L54

File: setup.py [hunk 5]
```diff
@@ -141,7 +144,7 @@ def run_tests(self):
         'Operating System :: OS Independent',
         'Programming Language :: Python',
         'Topic :: Internet :: WWW/HTTP :: Dynamic Content',
-        'Topic :: Software Development :: Libraries :: Python Modules',
+        'Topic :: Software Development :: Libraries :: Pyth0on Modules',
```
Comment (jirikuncar) [inline setup.py]: typo

File: setup.py [hunk 6]
```diff
@@ -151,4 +154,5 @@ def run_tests(self):
         'Development Status :: 3 - Alpha',
     ],
     cmdclass={'test': PyTest},
+
```
Comment (jirikuncar) [inline setup.py]: unnecessary line

File: tests/test_invenio_celery.py [hunk 7]
```diff
@@ -31,6 +31,21 @@
 from invenio_celery import InvenioCelery
 
 
+def test_0_autodiscovery(app):
```
Comment (jirikuncar) [inline tests/test_invenio_celery.py]: please use more descriptive names of tests

File: tests/test_invenio_celery.py [hunk 8]
```diff
@@ -31,6 +31,21 @@
...
+    """Test discovery."""
+    ext = InvenioCelery(app, enable_autodiscovery=False)
+    assert "projectwithsharedtask.tasks.compute" not in ext.celery.tasks.keys()
+
+
+def test_1_autodiscovery(app):
```
Comment (jirikuncar) [inline tests/test_invenio_celery.py]: https://github.com/inveniosoftware/invenio-celery/pull/10/files#r43890902

File: setup.py [hunk 9]
```diff
@@ -43,6 +43,8 @@
     'pytest-cov>=1.8.0',
     'pytest-pep8>=1.0.6',
     'pytest>=2.8.0',
+    'projectwithsharedtask'
```
Comment (lnielsen) [inline setup.py]: Remove this one?

File: .travis.yml [hunk 10]
```diff
@@ -52,8 +52,8 @@ before_install:
   - "requirements-builder --level=dev --req requirements-devel.txt setup.py > .travis-devel-requirements.txt"
 
 install:
-  - "travis_retry pip install -r .travis-${REQUIREMENTS}-requirements.txt"
-  - "travis_retry pip install -e .[all]"
+  - "travis_retry pip install -r .travis-${REQUIREMENTS}-requirements.txt  --process-dependency-links --trusted-host github.com"
```
Comment (jirikuncar) [inline .travis.yml]: you should not need it if you use mocking as we talked about.

File: setup.py (hunk 9 above)
Comment (jirikuncar) [inline setup.py]: @bruno314 please remove the package from PyPI too (https://pypi.python.org/pypi/projectwithsharedtask/0.2.0.dev20150000)

**Comment by bruno314** [comment]
that entrypoint_name value could be hardwired, I just wasn't 100% sure we weren't going to change the name

File: .gitignore [hunk 11]
```diff
@@ -1,3 +1,6 @@
+#
+.idea/
+.gitignore
```
Comment (lnielsen) [inline .gitignore]: ``.gitignore`` should not be added here.

File: .gitignore [hunk 12]
```diff
@@ -1,3 +1,6 @@
+#
+.idea/
```
Comment (lnielsen) [inline .gitignore]: Please add this to cookiecutter template as well

File: invenio_celery/ext.py [hunk 13]
```diff
@@ -26,24 +26,39 @@
...
             self.init_app(app, **kwargs)
 
-    def init_app(self, app, assets=None, **kwargs):
+    def init_app(self, app, assets=None,
+                 entrypoint_name="invenio_celery.tasks", **kwargs):
         """Initialize application object."""
         self.init_config(app.config)
         self.celery = FlaskCeleryExt(app).celery
+
+        # this reduces signature complexity, we want False only for testing
```
Comment (lnielsen) [inline invenio_celery/ext.py]:
```python
if entrypoint_name:
    all_modules = []
    for item in pkg_resources.iter_entry_points(group=entrypoint_name):
        all_modules.append(item.module_name)
    self.celery.autodiscover_tasks(all_modules, related_name="", force=True)
```

If you want to disable entrypoint loading you simply use ``InvenioCelery(entrypoint_name=None)``. I don't think entry point loading without auto discovery makes sense.

File: setup.py [hunk 14]
```diff
@@ -70,6 +70,7 @@
 
 
 class PyTest(TestCommand):
+
```
Comment (lnielsen) [inline setup.py]: Please remove line (you need to use PEP257 0.7.0)

File: tests/test_invenio_celery.py [hunk 15]
```diff
@@ -26,11 +26,40 @@
 
 from __future__ import absolute_import, print_function
 
-from mock import MagicMock
-
+from mock import MagicMock, patch
```
Comment (lnielsen) [inline tests/test_invenio_celery.py]: Please check import sorting.

File: invenio_celery/ext.py [hunk 16]
```diff
@@ -28,22 +28,34 @@
...
         """Initialize application object."""
         self.init_config(app.config)
         self.celery = FlaskCeleryExt(app).celery
+
+        all_modules = []
+        for item in pkg_resources.iter_entry_points(
+                group=entrypoint_name):
+            all_modules.append(item.module_name)
+
+        if entrypoint_name:
```
Comment (jirikuncar) [inline invenio_celery/ext.py]: `if all_modules:`

File: invenio_celery/ext.py [hunk 17]
```diff
@@ -28,22 +28,34 @@
...
 
-    def init_app(self, app, assets=None, **kwargs):
+    def init_app(self, app, assets=None,
+                 entrypoint_name="invenio_celery.tasks", **kwargs):
         """Initialize application object."""
         self.init_config(app.config)
         self.celery = FlaskCeleryExt(app).celery
+
+        all_modules = []
+        for item in pkg_resources.iter_entry_points(
```
Comment (jirikuncar) [inline invenio_celery/ext.py]: `if entrypoint_name:` + indent until line 58

**Comment by jirikuncar** [comment]
@bruno314 when you are done please remove WIP from the PR title.

File: .gitignore [hunk 18]
```diff
@@ -1,3 +1,5 @@
+#idea
```
Comment (lnielsen) [inline .gitignore]: Do you need both ``#idea`` and ``.idea``?

Comment (jirikuncar) [inline .gitignore]: If it's comment then it should be `# Idea product directory`

Comment (bruno314) [inline .gitignore]: Updated and added to cookiecutter's `.gitignore` as well

File: setup.py [hunk 19]
```diff
@@ -102,6 +102,7 @@ def run_tests(self):
         errno = pytest.main(self.pytest_args)
         sys.exit(errno)
 
+
```
Comment (jirikuncar) [inline setup.py]: is it necessary?

File: invenio_celery/ext.py [hunk 20]
```diff
@@ -28,22 +28,34 @@
...
         self.init_config(app.config)
         self.celery = FlaskCeleryExt(app).celery
+
+        if entrypoint_name:
+            all_modules = []
+            for item in pkg_resources.iter_entry_points(
+                    group=entrypoint_name):
+                all_modules.append(item.module_name)
+
+            self.celery.autodiscover_tasks(all_modules, related_name="",
```
Comment (jirikuncar) [inline invenio_celery/ext.py]: IMHO it's not necessary to call celery autodiscover tasks if `all_modules` is empty. @bruno314 please consider renaming `all_modules` to `task_packages`.

**Comment by lnielsen** [comment]
FYI. I'm creating a new release and will take care of merging the PR.

**Pull request closed without merging**